### PR TITLE
Update Spark CLI for `extraJavaOptions` and Fix SubmitBuilder tests

### PR DESCRIPTION
**Pull request by samvantran**
Resolves [DCOS-37965](https://jira.mesosphere.com/browse/DCOS-37965) and [DCOS-33686](https://jira.mesosphere.com/browse/DCOS-33686)
Readdresses PR #207 

This PR fixes CLI tests and allows Spark CLI to accept multiple extraJavaOptions when submitting a Spark job. It incorporates changes Nick made when he ported Spark onto the SDK as well as old commits from Art for long strings of Spark arguments.

### CODE REVIEW

File: cli/dcos-spark/main.go [hunk 1]
```diff
@@ -43,12 +43,7 @@ type SparkCommand struct {
 }
 
 func (cmd *SparkCommand) runSubmit(a *kingpin.Application, e *kingpin.ParseElement, c *kingpin.ParseContext) error {
-	marathonConfig, err := fetchMarathonConfig()
```
Comment (nickbp) [inline cli/dcos-spark/main.go]: FWIW it may be necessary to make dispatcher additions in order for this to be removable? I don't remember for sure.

Effectively the dispatcher could grab these values from its environment and include them in the task automatically, instead of requiring the CLI to pass them in from the outside.

File: cli/dcos-spark/submit_builder.go [hunk 2]
```diff
@@ -336,12 +347,36 @@ ARGLOOP:
...
+			// arg =  --conf
+			arg = strings.TrimPrefix(arg, "'")
+			next := args[i + 1]
+			if strings.HasPrefix(next, "'") {  // e.g. --driver-java-options '-Djava.config=setting... <-- next
+				inQuotes = true
+			}
+			next = strings.Replace(next, "'", "", -1)  // remove internal quotes
+			argsEquals = append(argsEquals, arg + "=" + next)
 			i += 2
+		} else if strings.HasSuffix(arg, "'") {  // attach the final arg to the string of args without the quote
```
Comment (elezar) [inline cli/dcos-spark/submit_builder.go]: What about escaped quotes? `\'`?

File: cli/dcos-spark/submit_builder.go [hunk 3]
```diff
@@ -336,12 +347,36 @@ ARGLOOP:
...
 			}
-			// merge this --key against the following val to get --key=val
-			argsEquals = append(argsEquals, arg+"="+args[i+1])
+
+			// if this is the beginning of a string of args e.g. '-Djava.option=setting -Djava.paramter=nonsense'
+			// we want to remove the leading single quote. Also remove internal quotes when the arg == --conf or some
+			// other named configuration
+			// e.g.: next = spark.driver.extraJavaOptions='-Djava.something=somethingelse
+			// arg =  --conf
+			arg = strings.TrimPrefix(arg, "'")
```
Comment (elezar) [inline cli/dcos-spark/submit_builder.go]: I don't quite follow why we need to trim the prefix here and then check why `next` also has the prefix.

File: cli/dcos-spark/submit_builder.go (hunk 2 above)
Comment (elezar) [reply cli/dcos-spark/submit_builder.go]: I just added the test:
```golang
	   java_options = []string{"-Djava.thirdConfig=\\'thirdSetting\\'"}
	   inputArgs = "--driver-java-option='-Djava.thirdConfig=\\'thirdSetting\\'' --conf spark.cores.max=8"
	   suite.testLongArgInternal(inputArgs, java_options)
```
and get the following stacktrace:
```
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 52 [running]:
testing.tRunner.func1(0xc4200780d0)
	/usr/local/go/src/testing/testing.go:622 +0x29d
panic(0x1423780, 0x169ff80)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
_/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark.cleanUpSubmitArgs(0x14a744a, 0x53, 0xc4204a6b28, 0x1, 0x1, 0xc420055958, 0x10b70be, 0x16a0710, 0x0, 0x0, ...)
	/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark/submit_builder.go:367 +0x3d8
_/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark.(*CliTestSuite).testLongArgInternal(0xc420195260, 0x14a744a, 0x53, 0xc420055a78, 0x1, 0x1)
	/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark/submit_builder_test.go:80 +0x7e
_/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark.(*CliTestSuite).TestStringLongArgs(0xc420195260)
	/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark/submit_builder_test.go:113 +0x333
reflect.Value.call(0xc4204b4720, 0xc4204a6260, 0x13, 0x148fb5d, 0x4, 0xc420045f80, 0x1, 0x1, 0xc420032ec0, 0x148e620, ...)
	/usr/local/go/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0xc4204b4720, 0xc4204a6260, 0x13, 0xc420032f80, 0x1, 0x1, 0x13d46c6, 0x12, 0x0)
	/usr/local/go/src/reflect/value.go:302 +0xa4
github.com/stretchr/testify/suite.Run.func2(0xc4200780d0)
	/Users/elezar/go/src/github.com/stretchr/testify/suite/suite.go:102 +0x25f
testing.tRunner(0xc4200780d0, 0xc420380e00)
	/usr/local/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:697 +0x2ca
--driver-java-option='-Djava.thirdConfig=\'thirdSetting\'' --conf spark.cores.max=8exit status 2
FAIL	_/Users/elezar/src/mesosphere/spark-build/cli/dcos-spark	0.029s
Error: Tests failed.
```

Comment (elezar) [reply cli/dcos-spark/submit_builder.go]: Adding the test:
```golang
inputArgs := "--driver-java-option='-Djava.thirdConfig=assetting --conf spark.cores.max=8"
java_options := []string{"-Djava.thirdConfig=thirdSetting"}
suite.testLongArgInternal(inputArgs, java_options)
```
Yields:
```
--- FAIL: TestCliTestSuite (0.00s)
    --- FAIL: TestCliTestSuite/TestIncorrectStrings (0.00s)
    	submit_builder_test.go:87: Expected to find -Djava.thirdConfig=thirdSetting at index 0. Found --driver-java-option=-Djava.thirdConfig=assetting instead
```
I would expect a relevant error code instead.

Comment (elezar) [reply cli/dcos-spark/submit_builder.go]: The following test case:
```golang
inputArgs := "--driver-java-option='-Djava.thirdConfig=\"a setting with a space\"' --conf spark.cores.max=8"
javaOptions := []string{"-Djava.thirdConfig=\"a setting with a space\""}
suite.testLongArgInternal(inputArgs, javaOptions)
```
Yields:
```
--- FAIL: TestCliTestSuite (0.00s)
    --- FAIL: TestCliTestSuite/TestIncorrectStrings (0.00s)
    	submit_builder_test.go:82: Failed to parse --driver-java-option='-Djava.thirdConfig="a setting with a space"' --conf spark.cores.max=8, should have 2 args, got 5
    	submit_builder_test.go:87: Expected to find -Djava.thirdConfig="a setting with a space" at index 0. Found --driver-java-option=-Djava.thirdConfig="a instead
```
I would expect the spaces in the argument to be maintained.

**Comment by samvantran** [comment]
@elezar I'm no go expert so this PR may actually be beyond me right now. I made an assumption that the approved PR #207 was fine as is and I was just combining it with some of Nick's changes when he migrated Spark to SDK to fix the CLI.

fwiw, when I ran `go test` I did see a pass
```
╭─  ~/go/src/github.com/mesosphere/spark-build/cli/dcos-spark                                                                                                                                           27.89    master   3   09:32:48
╰ go test -v
=== RUN   TestCliTestSuite
=== RUN   TestCliTestSuite/TestCleanUpSubmitArgs
=== RUN   TestCliTestSuite/TestPayloadCustomImageNoExecutor
=== RUN   TestCliTestSuite/TestPayloadSimple
Enabling forcePullImage by default. To disable this, set spark.mesos.executor.docker.forcePullImage=false
=== RUN   TestCliTestSuite/TestPayloadWithSecret
Enabling forcePullImage by default. To disable this, set spark.mesos.executor.docker.forcePullImage=false
Using Kerberos principal 'client@local'
Enabling forcePullImage by default. To disable this, set spark.mesos.executor.docker.forcePullImage=false
Using Kerberos principal 'client@local'
=== RUN   TestCliTestSuite/TestSaslSecret
Enabling forcePullImage by default. To disable this, set spark.mesos.executor.docker.forcePullImage=false
=== RUN   TestCliTestSuite/TestScoptAppArgs
=== RUN   TestCliTestSuite/TestStringLongArgs
--- PASS: TestCliTestSuite (0.00s)
    --- PASS: TestCliTestSuite/TestCleanUpSubmitArgs (0.00s)
    --- PASS: TestCliTestSuite/TestPayloadCustomImageNoExecutor (0.00s)
    --- PASS: TestCliTestSuite/TestPayloadSimple (0.00s)
    --- PASS: TestCliTestSuite/TestPayloadWithSecret (0.00s)
    --- PASS: TestCliTestSuite/TestSaslSecret (0.00s)
    --- PASS: TestCliTestSuite/TestScoptAppArgs (0.00s)
    --- PASS: TestCliTestSuite/TestStringLongArgs (0.00s)
PASS
ok  	github.com/mesosphere/spark-build/cli/dcos-spark	0.023s
```
But wrt to @susanxhuynh's question about this actually working when submitting spark jobs, I found it does not. I'm seeing the 2nd, 3rd, n extraJavaOption be passed as application arguments or the URI fetcher failing.

So in short:
- I agree, we should break out the fix SubmitBuilder into a separate PR. That would actually fix the two other PRs #371 and #372 bc they failed during backport.
- Address the comments you made specifically for the extraJavaOptions fixes

File: cli/dcos-spark/submit_builder.go [hunk 4]
```diff
@@ -336,12 +347,36 @@ ARGLOOP:
...
+			cleanedArg := strings.Replace(arg, "'", "", -1)
+			if inQuotes { // join this arg to the last one because it's all in quotes
+				argsEquals[len(argsEquals) - 1] = argsEquals[len(argsEquals) - 1] + " " + cleanedArg
+			} else {
+				if strings.Contains(arg, "'") {  // e.g. --driver-java-options='-Djava.firstConfig=firstSetting
+					inQuotes = true
+				}
+				// already joined or at the end, pass through
+				argsEquals = append(argsEquals, cleanedArg)
+			}
```
Comment (susanxhuynh) [inline cli/dcos-spark/submit_builder.go]: Maybe we can document somewhere that we expect single quotes to be used when a value has spaces in it.

File: cli/dcos-spark/submit_builder.go [hunk 5]
```diff
@@ -483,59 +492,30 @@ func buildSubmitJson(cmd *SparkCommand, marathonConfig map[string]interface{}) (
...
-			dispatcher_image, err := getStringFromTree(marathonConfig, []string{"app", "container", "docker", "image"})
-			if err != nil {
-				return "", err
-			}
-			args.properties["spark.mesos.executor.docker.image"] = dispatcher_image
-			imageSource = "dispatcher: container.docker.image"
-		} else {
-			args.properties["spark.mesos.executor.docker.image"] = cmd.submitDockerImage
-			imageSource = "flag: --docker-image"
-		}
```
Comment (susanxhuynh) [inline cli/dcos-spark/submit_builder.go]: I think removing this section will change the existing behavior: if a special Docker image was provided when installing the Dispatcher, that same image will also be used for Drivers and/or Executors (see the logic below). Unless we make some changes in the Dispatcher.

File: cli/dcos-spark/submit_builder_test.go [hunk 6]
```diff
@@ -71,6 +75,39 @@ func (suite *CliTestSuite) TestScoptAppArgs() {
 	}
 }
 
+func (suite *CliTestSuite) testLongArgInternal(inputArgs string, expectedArgs []string) {
```
Comment (susanxhuynh) [inline cli/dcos-spark/submit_builder_test.go]: Maybe "Helper" is a better name for this function rather than "Internal".

File: cli/dcos-spark/submit_builder_test.go [hunk 7]
```diff
@@ -71,6 +75,39 @@ func (suite *CliTestSuite) TestScoptAppArgs() {
...
+       suite.testLongArgInternal(inputArgs, java_options)
+       inputArgs = "--conf spark.driver.extraJavaOptions='-Djava.firstConfig=firstSetting -Djava.secondConfig=secondSetting' --conf spark.cores.max=8"
+       suite.testLongArgInternal(inputArgs, java_options)
+       inputArgs = "--executor-java-options '-Djava.firstConfig=firstSetting -Djava.secondConfig=secondSetting' --conf spark.cores.max=8"
+       suite.testLongArgInternal(inputArgs, java_options)
+       inputArgs = "--conf spark.executor.extraJavaOptions='-Djava.firstConfig=firstSetting -Djava.secondConfig=secondSetting' --conf spark.cores.max=8"
+       suite.testLongArgInternal(inputArgs, java_options)
+
+       java_options = append(java_options, "-Djava.thirdConfig=thirdSetting")
+       inputArgs = "--driver-java-option='-Djava.firstConfig=firstSetting -Djava.secondConfig=secondSetting -Djava.thirdConfig=thirdSetting' --conf spark.cores.max=8"
```
Comment (susanxhuynh) [inline cli/dcos-spark/submit_builder_test.go]: Typo in "--driver-java-option" (missing an "s")

File: Makefile [hunk 8]
```diff
@@ -23,6 +23,8 @@ HADOOP_VERSION ?= $(shell jq ".default_spark_dist.hadoop_version" "$(ROOT_DIR)/m
 SPARK_DIR ?= $(ROOT_DIR)/spark
 $(SPARK_DIR):
 	git clone $(SPARK_REPO_URL) $(SPARK_DIR)
+	cd $(SPARK_DIR)
```
[the remark this answers is not in the document]
Comment (samvantran) [reply Makefile]: ooh sorry this was an accident! Should not have been there.

**Comment by samvantran** [comment]
Closing this PR to separate fixing tests and adding functionality
